src/hotspot/share/prims/jvmtiThreadState.inline.hpp

```diff
@@ -78,17 +78,30 @@ inline JvmtiThreadState* JvmtiThreadState::state_for_while_locked(JavaThread *th
 
   NoSafepointVerifier nsv;  // oop is safe to use.
 
-  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
-    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
-  }
-
   // In a case of unmounted virtual thread the thread can be null.
   JvmtiThreadState *state = thread == nullptr ? nullptr : thread->jvmti_thread_state();
 
-  if (state == nullptr && thread != nullptr && thread->is_exiting()) {
-    // Don't add a JvmtiThreadState to a thread that is exiting.
+  if (state == nullptr && thread != nullptr &&
+      (thread->is_exiting() || thread->is_attaching_via_jni())) {
+    // Don't add a JvmtiThreadState to a thread that is exiting or is attaching.
+    // When a thread is attaching, it may not have a Java level thread object
+    // created yet.
     return nullptr;
   }
+
+#ifdef ASSERT
+  // Make sure we don't see an incomplete state. An incomplete state can cause
+  // a duplicate JvmtiThreadState being created below and bound to the 'thread'
+  // incorrectly, which leads to stale JavaThread* from the JvmtiThreadState
+  // after the thread exits.
+  if (state != nullptr) {
+    assert(state->get_thread_oop() != nullptr, "incomplete state");
+  }
+#endif
```

Comment:
Nit: I would suggest to write this assert in the form:
```
  // Make sure we don't see an incomplete state. An incomplete state can cause
  // a duplicate JvmtiThreadState being created below and bound to the 'thread'
  // incorrectly, which leads to stale JavaThread* from the JvmtiThreadState
  // after the thread exits.
   assert(state == nullptr || state->get_thread_oop() != nullptr, "incomplete state");
```
The `#ifdef ASSERT` and `#endif` are not needed then.

Reply:
Changed as suggested.